src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -342,6 +343,17 @@ public async ValueTask DisposeAsync()
     {
         if (Interlocked.Exchange(ref _state, StateDisposed) is not StateDisposed)
         {
+            // Complete outstanding subscriptions.
+            lock (_lock)
+            {
+                foreach (var item in _outgoingChannels)
+                {
+                    item.Writer.Complete();
```

Comment:
Completion still requires consumers to process any pending messages in the channel. The intent of the original design here was to have consumers cancel immediately and ignore any unprocessed messages.

Reply:
You're mixing the reader side and writer side into one thing. They're separate.

The code here doesn't stop the reader side of the channel from choosing to cancel reading. That's still a thing.

Completing the writer is purely additive. By completing the writer on dispose, any readers that haven't canceled reading will exit once they read to the end. It's safer and prevents hanging code. It also makes unit testing easier as discussed [here](https://github.com/dotnet/aspire/pull/1348#discussion_r1428707324).

Reply:
Not mixing anything. We don't need completion when we have cancellation for consumers. The assumption through the design was that cancellation would clean up everything neatly. Gaps in this were identified for me to fix post-merge (done in #1589).

Introducing completion _and_ disposal introduces multiple ways for us to have to think about cleaning things up, which increases the chances for bugs. I'd rather have a single approach, which would mean using cancellation tokens as they're non-negotiable in async code, and are composable.

Reply:
The producer should complete the channel writer when it is shutting down. I don't know if anyone in MS has written best practice around using channels, but if they did then I'm sure that would be in it.

Example in System.Net.Quic: https://github.com/dotnet/runtime/blob/abffa8fd7c9a9eaaaba70ca8052d605b85918bd4/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs#L417
When the QUIC connection is disposed it completes the pending connection channel.

Example in HttpClient:
https://github.com/dotnet/runtime/blob/6495c22ec784821269ae27f933da4f5074127192/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1855

We don't use channels much in ASP.NET Core, but we use pipelines a lot, and it is another producer/consumer. We always complete the writer when shutting down:
https://github.com/dotnet/aspnetcore/blob/7b9e6b06b116b4c3bebfbcb0a2f82fe79c907363/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs#L103-L104
(there are two pipes for the bidirectional nature of the socket)

Reply:
There aren't channel best practices but I found pipelines best practices. They do recommend always completing the reader/writer: https://learn.microsoft.com/en-us/dotnet/standard/io/pipelines#tips-for-using-pipereader-and-pipewriter

Reply:
There are certainly cases where flushing pending work is desirable, and completion is an elegant way to do so. In this case we are eagerly cancelling the channel reader, so completion has no observable effect. That was a conscious choice in writing the code here.

Reply:
I am not a fan of the coding style where we fire tasks and don't await their shutdown and cleanup. Having written this sort of async code for years now, it's a pattern that generally leads to hard to track down leaks. ASP.NET Core attempts to store and wait on tasks in most cases, and in the cases where they are ignored there's a giant comment because it's the exception, not the rule.